cmd/ipfs/migration.go

```diff
@@ -0,0 +1,292 @@
+	} else {
+		peers = peeringCfg.Peering.Peers
+	}
+
+	return
+}
+
+// getMigrationFetcher creates one or more fetchers according to
+// config.Migration.DownloadSources.  If an IpfsFetcher is required, then
+// bootstrap and peer information in read from the config file in repoRoot,
+// unless repoRoot is nil.
+func getMigrationFetcher(cfg *config.Migration, repoRoot *string) (migrations.Fetcher, error) {
+	const httpUserAgent = "go-ipfs"
+
+	// Fetch migrations from current distribution, or location from environ
+	fetchDistPath := migrations.GetDistPathEnv(migrations.CurrentIpfsDist)
+
+	var fetchers []migrations.Fetcher
+	for _, src := range cfg.DownloadSources {
+		src := strings.TrimSpace(src)
+		switch src {
+		case "IPFS", "ipfs":
+			bootstrap, peers := readIpfsConfig(repoRoot)
+			fetchers = append(fetchers, ipfsfetcher.NewIpfsFetcher(fetchDistPath, 0, bootstrap, peers))
+		case "HTTPS", "https", "HTTP", "http":
+			fetchers = append(fetchers, migrations.NewHttpFetcher(fetchDistPath, "", httpUserAgent, 0))
+		default:
+			u, err := url.Parse(src)
+			if err != nil {
+				return nil, fmt.Errorf("bad gateway address: %s", err)
+			}
+			switch u.Scheme {
+			case "":
+				u.Scheme = "https"
+			case "https", "http":
+			default:
+				return nil, errors.New("bad gateway address: url scheme must be http or https")
+			}
+			fetchers = append(fetchers, migrations.NewHttpFetcher(fetchDistPath, u.String(), httpUserAgent, 0))
+		case "":
+			// Ignore empty string
+		}
+	}
+	if len(fetchers) == 0 {
+		return nil, errors.New("no sources specified")
+	}
+
+	if len(fetchers) == 1 {
+		return fetchers[0], nil
+	}
+
+	// Wrap fetchers in a MultiFetcher to try them in order
+	return migrations.NewMultiFetcher(fetchers...), nil
+}
+
+func addMigrations(ctx context.Context, node *core.IpfsNode, fetcher migrations.Fetcher, pin bool) error {
+	if mf, ok := fetcher.(*migrations.MultiFetcher); ok {
+		fetcher = mf.LastUsed()
+	}
+
+	switch f := fetcher.(type) {
+	case *ipfsfetcher.IpfsFetcher:
+		// Add migrations by connecting to temp node and getting from IPFS
+		err := addMigrationPaths(ctx, node, f.AddrInfo(), f.FetchedPaths(), pin)
+		if err != nil {
+			return err
+		}
+	case *migrations.HttpFetcher:
+		// Add the downloaded migration files directly
+		if migrations.DownloadDirectory != "" {
+			var paths []string
+			err := filepath.Walk(migrations.DownloadDirectory, func(filePath string, info os.FileInfo, err error) error {
+				if info.IsDir() {
+					return nil
+				}
+				paths = append(paths, filePath)
+				return nil
+			})
+			if err != nil {
+				return err
+			}
+			err = addMigrationFiles(ctx, node, paths, pin)
+			if err != nil {
+				return err
+			}
+		}
+	default:
+		return errors.New("Cannot get migrations from unknown fetcher type")
+	}
+
+	return nil
+}
+
+// addMigrationFiles adds the files at paths to IPFS, optionally pinning them
+func addMigrationFiles(ctx context.Context, node *core.IpfsNode, paths []string, pin bool) error {
```

Comment:
Given that we just merged CAR export on gateways (https://github.com/ipfs/go-ipfs/pull/8111) I feel much less strongly about this and think we can leave it as it is.

What we can do is in a later RC (or next release) switch the HTTP Fetcher to really be a CAR fetcher and then this problem goes away. For the time being we control how the migrations are created so we can just leave the defaults.

cc @Stebalien as a heads up

Reply:
Let me know if you want me to remove the support for adding downloads from HTTP to IPFS.  ...since the CAR download is on the horizon, and less code to worry about.